common/dbconfig.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright 2021-present Pinterest. All Rights Reserved.
+//
+// @author Premkumar (prem@pinterest.com)
+//
+
+#include "common/dbconfig.h"
+#include "common/segment_utils.h"
+#include <utility>
+
+#include <gflags/gflags.h>
+#include <glog/logging.h>
+
+DECLARE_int32(replicator_replication_mode);
+
+namespace common {
+
+DBConfigManager* DBConfigManager::get() {
+  static DBConfigManager instance;
+  return &instance;
+}
+
+DBConfigManager::DBConfigManager() : dbConfig_(new DBConfig()), fDataLoaded_(false) {
+}
+
+bool DBConfigManager::loadJsonStr(const std::string& content) {
+  Json::Value root;
+  Json::Reader reader;
+  if (!reader.parse(content, root) || !root.isObject()) {
+    LOG(ERROR) << "Could not load config";
+    return false;
+  }
+
+  return loadJsonObject(root);
+}
+
+bool DBConfigManager::loadJsonObject(const Json::Value& root) {
+  try {
+    auto newConfig = parseConfig(root);
+    if (newConfig == nullptr) {
+      return false;
+    }
+
+    LOG(INFO) << "Successfully updated config";
+    std::atomic_exchange_explicit(&dbConfig_, newConfig, std::memory_order_release);
+    fDataLoaded_.store(true);
+    return true;
+  } catch (std::exception& ex) {
+    LOG(ERROR) << "Exception from parseConfig: " << ex.what();
+    return false;
+  }
+}
+
+DBConfigPtr DBConfigManager::parseConfig(const Json::Value& root) {
+  static const std::string ACK_MODE = "ack_mode";
+  static const std::string DATASET = "dataset";
+
+  DBConfig newDBConfig;
+
+  if (root.isMember(DATASET)) {
```

Comment:
How about we reorganize the code like the following?

```
if (!root.isMember(DATASET)) {
 return nullptr;
}

// ...
```

Reply:
@premkumr can you please address this comment?

It makes the code easier to read and maintain.
Also, as in its current form, it can't differentiate between an invalid config and a valid config with empty config content (use the default settings). In another word, if someone wants to change to use the default settings by submitting an empty config, the new config will be rejected.

Reply:
My bad on missing this comment. Yes it cannot differentiate between an empty config and and invalid config. I wanted to ignore an empty config. missing or invalid will be caught by exception and ignored. .. If we want to accept an empty config, effectively '{}' , I can change it in PR:565 .  Let me know @newpoo ...